File: examples/student_grades.py
"""Student grade management system."""
import json
import statistics

class Student:
    """Represents a student with grades."""

    def __init__(self, name, grades=None):
        self.name = name
        self.grades = grades or []

    def add_grade(self, grade):
        if 0 <= grade <= 100:
            self.grades.append(grade)
            return True
        return False

    def get_average(self):
        if not self.grades:
            return 0
        return sum(self.grades) / len(self.grades)

    def get_letter_grade(self):
        avg = self.get_average()
        if avg >= 90:
            return 'A'
        elif avg >= 80:
            return 'B'
        elif avg >= 70:
            return 'C'
        elif avg >= 60:
            return 'D'
        return 'F'

class Classroom:
    """Manages multiple students."""

    def __init__(self, class_name):
        self.class_name = class_name
        self.students = []

    def add_student(self, student):
        self.students.append(student)

    def get_class_average(self):
        if not self.students:
            return 0
        total = 0
        for student in self.students:
            total += student.get_average()
        return total / len(self.students)

    def get_top_student(self):
        if not self.students:
            return None
        best = self.students[0]
        for student in self.students:
            if student.get_average() > best.get_average():
                best = student
        return best

    def save_to_file(self, filepath):
        data = {
            'class': self.class_name,
            'students': [
                {'name': s.name, 'grades': s.grades}
                for s in self.students
            ]
        }
        with open(filepath, 'w') as f:
            json.dump(data, f, indent=2)

def main():
    classroom = Classroom("Math 101")

    alice = Student("Alice")
    alice.add_grade(85)
    alice.add_grade(92)
    alice.add_grade(78)
    classroom.add_student(alice)

    bob = Student("Bob")
    bob.add_grade(90)
    bob.add_grade(88)
    bob.add_grade(95)
    classroom.add_student(bob)

    print(f"Class: {classroom.class_name}")
    print(f"Class Average: {classroom.get_class_average():.1f}")

    top = classroom.get_top_student()
    if top:
        print(f"Top Student: {top.name} ({top.get_average():.1f})")

    classroom.save_to_file("grades.json")

if __name__ == '__main__':
    main()

File: examples/student_grades.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Student structure
struct Student {
    char name[50];
    int grades[10];
    int grade_count;
};

// Calculate average grade
float get_average(struct Student *s) {
    if (s->grade_count == 0) return 0.0;
    int sum = 0;
    for (int i = 0; i < s->grade_count; i++) {
        sum += s->grades[i];
    }
    return (float)sum / s->grade_count;
}

// Get letter grade
char get_letter_grade(float avg) {
    if (avg >= 90) return 'A';
    if (avg >= 80) return 'B';
    if (avg >= 70) return 'C';
    if (avg >= 60) return 'D';
    return 'F';
}

// Add grade to student
int add_grade(struct Student *s, int grade) {
    if (grade < 0 || grade > 100) return 0;
    if (s->grade_count >= 10) return 0;
    s->grades[s->grade_count] = grade;
    s->grade_count++;
    return 1;
}

int main() {
    struct Student alice;
    strcpy(alice.name, "Alice");
    alice.grade_count = 0;

    add_grade(&alice, 85);
    add_grade(&alice, 92);
    add_grade(&alice, 78);

    float avg = get_average(&alice);
    char letter = get_letter_grade(avg);

    printf("Student: %s\n", alice.name);
    printf("Average: %.1f\n", avg);
    printf("Grade: %c\n", letter);

    return 0;
}
